src/output-flex.cpp

```diff
@@ -1055,33 +1043,129 @@ void output_flex_t::pending_way(osmid_t id)
+    }
+
+    // We have established that we have a table. Get the 'ways' field...
+    lua_getfield(lua_state(), -1, "ways");
+    int const ltype = lua_type(lua_state(), -1);
+
+    // No 'ways' field, that is okay, nothing to be marked.
+    if (ltype == LUA_TNIL) {
+        lua_pop(lua_state(), 2); // return value (a table), ways field (nil)
+        return;
+    }
+
+    if (ltype != LUA_TTABLE) {
+        throw std::runtime_error{
+            "Table returned from select_relation_members() contains 'ways' "
+            "field, but it isn't an array table"};
+    }
+
+    // Iterate over the 'ways' table to get all ids...
+    lua_pushnil(lua_state());
+    while (lua_next(lua_state(), -2) != 0) {
+        if (!lua_isnumber(lua_state(), -2)) {
+            throw std::runtime_error{
+                "Table returned from select_relation_members() contains 'ways' "
+                "field, but it isn't an array table"};
+        }
+
+        osmid_t const id = lua_tointeger(lua_state(), -1);
+        if (id == 0) {
+            throw std::runtime_error{
+                "Table returned from select_relation_members() contains 'ways' "
+                "field, which must contain an array of non-zero integer way "
+                "ids"};
+        }
+
+        m_stage2_way_ids->set(id);
+        lua_pop(lua_state(), 1); // value pushed by lua_next()
+    }
+
+    lua_pop(lua_state(), 2); // return value (a table), ways field (a table)
+}
+
+void output_flex_t::select_relation_members(osmid_t id)
+{
+    if (!m_select_relation_members) {
+        return;
+    }
+
+    if (!m_mid->relation_get(id, m_rels_buffer)) {
+        return;
+    }
+
+    select_relation_members(m_rels_buffer.get<osmium::Relation>(0));
+
+    m_rels_buffer.clear();
+}
+
 void output_flex_t::pending_relation(osmid_t id)
 {
-    if (!m_process_relation) {
+    if (!m_process_relation && !m_select_relation_members) {
         return;
     }
 
-    // Try to fetch the relation from the DB
-    // Note that we cannot use the global buffer here because
-    // we cannot keep a reference to the relation and an autogrow buffer
-    // might be relocated when more data is added.
     if (!m_mid->relation_get(id, m_rels_buffer)) {
         return;
     }
+    auto const &relation = m_rels_buffer.get<osmium::Relation>(0);
 
+    select_relation_members(relation);
     delete_from_tables(osmium::item_type::relation, id);
 
+    if (m_process_relation) {
+        m_context_relation = &relation;
+        get_mutex_and_call_lua_function(m_process_relation, relation);
+        m_context_relation = nullptr;
+    }
+
+    m_rels_buffer.clear();
+}
+
+void output_flex_t::pending_relation_stage1c(osmid_t id)
+{
+    if (!m_process_relation) {
+        return;
+    }
+
+    if (!m_mid->relation_get(id, m_rels_buffer)) {
+        return;
+    }
     auto const &relation = m_rels_buffer.get<osmium::Relation>(0);
 
+    m_disable_add_row = true;
```

Comment:
Shouldn' t that be set to false again after the function returns? I' d expect that add_row() now does nothing in stage2 as well (although that is obviously wrong because the tests pass).

Reply:
Threading confusion strikes again. This is only set in the sub-threads, while the stage2 runs in the main thread. I have changed the code anyway and disabled it after the call again, to avoid this confusion and make it more robust if we decide to change in what thread things run.